Session time zones (#44149)
ClickHouse/ClickHouse: docs/en/operations/settings/settings.md

Use this setting only for backward compatibility if your use cases depend on old syntax.
:::

## session_timezone {#session_timezone}

> If specified, sets an implicit timezone (instead of server-default). All DateTime/DateTime64 values (and/or functions results) that have no explicit timezone specified are treated as having this timezone instead of default.

Reviewer: Setting this to `''` (empty string), effectively resets implicit timezone to server timezone.

Examples:

```clickhouse
SELECT timeZone(), serverTimezone() FORMAT TSV

Europe/Berlin	Europe/Berlin
```

```clickhouse
SELECT timeZone(), serverTimezone() SETTINGS session_timezone = 'Asia/Novosibirsk' FORMAT TSV

Asia/Novosibirsk	Europe/Berlin
```

```clickhouse
SELECT toDateTime64(toDateTime64('1999-12-12 23:23:23.123', 3), 3, 'Europe/Zurich') SETTINGS session_timezone = 'America/Denver' FORMAT TSV

1999-12-13 07:23:23.123
```

Possible values:

> -    Any valid timezone in `Region/Place` notation, e.g. `Europe/Berlin`

Reviewer: ```suggestion
-    Any timezone name from `system.time_zones`, e.g. `Europe/Berlin`, `UTC` or `Zulu`
```

Use this setting only for backward compatibility if your use cases depend on old syntax.
:::

## session_timezone {#session_timezone}

> If specified, sets an implicit timezone (instead of server-default). All DateTime/DateTime64 values (and/or functions results) that have no explicit timezone specified are treated as having this timezone instead of default.

Author: Why? I think it is a good one. It is really implicit e.g. in `toDateTime('2022-12-12 12:12:12.11', 2)`, it is not set explicitly

Use this setting only for backward compatibility if your use cases depend on old syntax.
:::

> ## session_timezone {#session_timezone}

Author: there is already a setting called `timezone` in `config.xml`. This may also introduce some mess. 
Also, don't remember exactly why, but I **_had_** to change the name, I used to call it `timezone`, but there were some errors. Maybe we can come up with a better naming option :) 
maybe something like `user_defined_timezone` or `user_timezone` would be fine

> ## session_timezone {#session_timezone}

Reviewer 2: Yes, setting `timezone` in "config.xml" *is* the global/server-level timezone.

As far as I see, there is no precedence for an equally named global/session setting, but I don't think anything can mess up. It would be very appreciated if you could make a try to rename `session_timezone` to `timezone` in a separate PR, after this one is merged.

(`user_defined_timezone` / `user_timezone` ar a bit too chatty for my taste)

> ## session_timezone {#session_timezone}

Author: yes, agree

> ## session_timezone {#session_timezone}

Author: It actually would be nicer to rename server's setting to `server_timezone`, but I guess it is a really bad option

> ## session_timezone {#session_timezone}

Reviewer 2: It is a bad idea :-) (and unfortunately settings can't be aliased as of now, so we can't deprecate setting "timezone").

> ## session_timezone {#session_timezone}

Author: Ok, after this is merged I can try to change it to `timezone`

If specified, sets an implicit timezone (instead of [server default](../server-configuration-parameters/settings.md#server_configuration_parameters-timezone).
All DateTime/DateTime64 values (and/or functions results) that have no explicit timezone specified are treated as having this timezone instead of default.

> A value of `''` (empty string) configures the session timezone to the server default timezone.

Author: Well, checking if string is empty is cheap, easy and obvious, and it also doesn't break the way things currently work, while empty string is passed by default, e.g. in TimezoneMixin and all its children.
This would introduce additional checks and string comparisons here and there, not good for performance also.
Moreover, in all other settings emptiness means that setting is not set and default behavior is working.
IMO when one wants to change back to default, empty string looks more default than some string with empty value, especially if he has used other settings previously

> A value of `''` (empty string) configures the session timezone to the server default timezone.

Reviewer 2: Makes sense, thanks.

Examples:

```sql

> SELECT timeZone(), serverTimezone() FORMAT TSV

Author: added

Europe/Berlin	Europe/Berlin
```

```sql
SELECT timeZone(), serverTimezone() SETTINGS session_timezone = 'Asia/Novosibirsk' FORMAT TSV

Asia/Novosibirsk	Europe/Berlin
```

```sql
SELECT toDateTime64(toDateTime64('1999-12-12 23:23:23.123', 3), 3, 'Europe/Zurich') SETTINGS session_timezone = 'America/Denver' FORMAT TSV

1999-12-13 07:23:23.123
```

:::warning
The way this setting affects parsing of Date or DateTime types may seem non-obvious, see example and explanation below:
:::

```sql
CREATE TABLE test_tz (`d` DateTime('UTC')) ENGINE = Memory AS SELECT toDateTime('2000-01-01 00:00:00', 'UTC');

SELECT *, timezone() FROM test_tz WHERE d = toDateTime('2000-01-01 00:00:00') SETTINGS session_timezone = 'Asia/Novosibirsk'

> 0 rows in set.

Author: Yes, exactly like this
